rpc/tpch_utils.py

```diff
@@ -0,0 +1,170 @@
+# Code adopted from decima-sim to use their profiles of TPC-H queries
+
+import ast
+import json
+import os
+from typing import Mapping, Sequence
+
+import networkx as nx
+import numpy as np
+
+HOME_TPCH_DIR = "../profiles/workload/tpch_decima/"
+TPCH_SUBDIR = "2g/"
+
+
+class SetWithCount(object):
+    """
+    allow duplication in set
+    """
+
+    def __init__(self):
+        self.set = {}
+
+    def __contains__(self, item):
+        return item in self.set
+
+    def add(self, item):
+        if item in self.set:
+            self.set[item] += 1
+        else:
+            self.set[item] = 1
+
+    def clear(self):
+        self.set.clear()
+
+    def remove(self, item):
+        self.set[item] -= 1
+        if self.set[item] == 0:
+            del self.set[item]
+
+
+def pre_process_task_duration(task_duration):
+    # remove fresh durations from first wave
+    clean_first_wave = {}
+    for e in task_duration["first_wave"]:
+        clean_first_wave[e] = []
+        fresh_durations = SetWithCount()
+        # O(1) access
+        for d in task_duration["fresh_durations"][e]:
+            fresh_durations.add(d)
+        for d in task_duration["first_wave"][e]:
+            if d not in fresh_durations:
+                clean_first_wave[e].append(d)
+            else:
+                # prevent duplicated fresh duration blocking first wave
+                fresh_durations.remove(d)
+
+
+def get_all_stage_info_for_query(query_num):
+    task_durations = np.load(
+        os.path.join(
+            HOME_TPCH_DIR, TPCH_SUBDIR, "task_duration_" + str(query_num) + ".npy"
```

Comment:
Should we get the `TPCH_SUBDIR` label from the application name?

We might want to run a mix of TPC-H queries from different scaling sizes.